Implement trusted-types-eval CSP script-src keyword

lukewarlow opened the pull request:
#### 003226eb1dbf7deec539b13f1fae3864fd4f7a31
<pre>
Implement trusted-types-eval CSP script-src keyword
<a href="https://bugs.webkit.org/show_bug.cgi?id=285603">https://bugs.webkit.org/show_bug.cgi?id=285603</a>

Reviewed by Ryan Reno.

Add a new script-src keyword that allows evaluating strings as JS,
similar to unsafe-eval but with the added requirement that trusted-types
must also be enabled and enforced.

* LayoutTests/http/tests/security/contentSecurityPolicy/block-eval-onload-in-nested-about-blank-iframe-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/eval-blocked-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/eval-blocked-in-about-blank-iframe-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/eval-blocked-in-external-script-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/eval-blocked-in-subframe-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/eval-scripts-setInterval-blocked-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/eval-scripts-setTimeout-blocked-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/function-constructor-blocked-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/module-eval-blocked-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/module-eval-blocked-in-external-script-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/worker-blob-inherits-csp-blocks-eval-expected.txt:
* LayoutTests/http/tests/security/isolatedWorld/image-load-should-not-bypass-main-world-csp-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/script-src/script-src-trusted_types_eval_with_report_only_require_trusted_types_eval-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/script-src/script-src-trusted_types_eval_with_report_only_require_trusted_types_eval.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/script-src/script-src-trusted_types_eval_with_report_only_require_trusted_types_eval.html.headers: Added.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/script-src/script-src-trusted_types_eval_with_require_trusted_types_eval-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/script-src/script-src-trusted_types_eval_with_require_trusted_types_eval.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/script-src/script-src-trusted_types_eval_with_require_trusted_types_eval.html.headers: Added.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/script-src/script-src-trusted_types_eval_without_require_trusted_types_eval-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/script-src/script-src-trusted_types_eval_without_require_trusted_types_eval.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/script-src/script-src-trusted_types_eval_without_require_trusted_types_eval.html.headers: Added.
* LayoutTests/security/contentSecurityPolicy/worker-inherits-blocks-eval-expected.txt:
* Source/JavaScriptCore/debugger/DebuggerEvalEnabler.h:
(JSC::DebuggerEvalEnabler::DebuggerEvalEnabler):
(JSC::DebuggerEvalEnabler::~DebuggerEvalEnabler):
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::eval):
* Source/JavaScriptCore/runtime/DirectEvalExecutable.cpp:
(JSC::DirectEvalExecutable::create):
* Source/JavaScriptCore/runtime/FunctionConstructor.cpp:
(JSC::constructFunction):
* Source/JavaScriptCore/runtime/IndirectEvalExecutable.cpp:
(JSC::IndirectEvalExecutable::createImpl):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::trustedTypesEnforcement const):
(JSC::JSGlobalObject::setTrustedTypesEnforcement):
(JSC::JSGlobalObject::requiresTrustedTypes const): Deleted.
(JSC::JSGlobalObject::setRequiresTrustedTypes): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/WebCore/bindings/js/ScriptController.cpp:
(WebCore::ScriptController::setTrustedTypesEnforcement):
(WebCore::ScriptController::setRequiresTrustedTypes): Deleted.
* Source/WebCore/bindings/js/ScriptController.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setTrustedTypesEnforcement):
(WebCore::Document::setRequiresTrustedTypes): Deleted.
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/EmptyScriptExecutionContext.h:
* Source/WebCore/dom/ScriptExecutionContext.h:
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::ContentSecurityPolicy::didCreateWindowProxy const):
(WebCore::ContentSecurityPolicy::applyPolicyToScriptExecutionContext):
(WebCore::ContentSecurityPolicy::allowEval const):
(WebCore::ContentSecurityPolicy::requireTrustedTypesForSinkGroup const):
* Source/WebCore/page/csp/ContentSecurityPolicy.h:
* Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.cpp:
(WebCore::checkTrustedEval):
(WebCore::ContentSecurityPolicyDirectiveList::create):
* Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.h:
(WebCore::ContentSecurityPolicyDirectiveList::trustedEvalEnabled const):
(WebCore::ContentSecurityPolicyDirectiveList::setTrustedEvalEnabled):
* Source/WebCore/page/csp/ContentSecurityPolicySourceList.cpp:
(WebCore::ContentSecurityPolicySourceList::parseSource):
* Source/WebCore/page/csp/ContentSecurityPolicySourceList.h:
(WebCore::ContentSecurityPolicySourceList::allowTrustedEval const):
* Source/WebCore/page/csp/ContentSecurityPolicySourceListDirective.h:
(WebCore::ContentSecurityPolicySourceListDirective::allowTrustedEval const):
* Source/WebCore/workers/WorkerOrWorkletGlobalScope.cpp:
(WebCore::WorkerOrWorkletGlobalScope::setTrustedTypesEnforcement):
(WebCore::WorkerOrWorkletGlobalScope::setRequiresTrustedTypes): Deleted.
* Source/WebCore/workers/WorkerOrWorkletGlobalScope.h:
* Source/WebCore/workers/WorkerOrWorkletScriptController.cpp:
(WebCore::WorkerOrWorkletScriptController::setTrustedTypesEnforcement):
(WebCore::WorkerOrWorkletScriptController::setRequiresTrustedTypes): Deleted.
* Source/WebCore/workers/WorkerOrWorkletScriptController.h:

Canonical link: <a href="https://commits.webkit.org/292217@main">https://commits.webkit.org/292217@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/928d5a16f4e0360d6039669aed978f77a574e2bf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95267 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/14867 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/4725 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/100311 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/45768 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/97315 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15155 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23298 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72650 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29927 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98270 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11317 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/85995 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52982 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11030 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/3732 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45107 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/87938 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81205 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/3827 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/102349 "Build is in progress. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; Compiled WebKit") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/93890 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/22315 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/16260 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/81648 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/22563 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82014 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81046 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20279 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25615 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3012 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/15568 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22285 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/27411 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/116578 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/21944 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/33313 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/25418 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/23683 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

lukewarlow commented on LayoutTests/imported/w3c/web-platform-tests/content-security-policy/script-src/script-src-trusted_types_eval_with_report_only_require_trusted_types_eval.html:
```diff
@@ -0,0 +1,103 @@
+    <!-- Report Only CSP served: require-trusted-types-for 'script' -->
+</head>
+
+<body>
```
Export these minus the 'webkit-test-runner' comments

lukewarlow replied:
https://github.com/web-platform-tests/wpt/pull/50458

lukewarlow commented on Source/JavaScriptCore/interpreter/Interpreter.cpp:
```diff
@@ -157,7 +157,7 @@ JSValue eval(CallFrame* callFrame, JSValue thisValue, JSScope* callerScopeChain,
 
     TopCallFrameSetter topCallFrame(vm, callFrame);
-    if (!globalObject->evalEnabled()) {
+    if (!globalObject->evalEnabled() && !(Options::useTrustedTypes() && globalObject->trustedEvalEnabled() && globalObject->requiresTrustedTypesEnforced())) {
```
Currently this is implemented under the same JSC option as the rest of the Trusted Types, need to decide if this should be behind it's own flag or not. (Ideally both would ship together though)

darinadler commented on Source/JavaScriptCore/runtime/JSGlobalObject.h:
```diff
@@ -587,8 +587,10 @@ class JSGlobalObject : public JSSegmentedVariableObject {
+    bool m_trustedEvalEnabled { true };
     bool m_webAssemblyEnabled { true };
     bool m_requiresTrustedTypes { true };
+    bool m_requiresTrustedTypesEnforced { true };
```
This seems like peculiar wording: "requires trusted types enforced" doesn’t seem like idiomatic language. Maybe we could find a way to word this more the way a person would say it in conversation, or in a way that matches what the specification calls it.

kmiller68 replied:
Yeah, these bools should be an `enum` that conveys the TrustedTypes enforcement status. It seems like it could also contain the m_trustedEvalEnabled flag too.

lukewarlow replied:
Okay yeah a three value enum for None, ReportOnly and Enforced makes sense. I'll update accordingly.

Not sure if including trusted types eval will work but I'll take a look when doing the change.

kmiller68 replied:
It seems like `trustedEvalEnabled == true` only matters when `requiresTrustedTypesEnforced == true` so it seems like you could just make a fourth enum value `EnforcedWithEvalEnabled`.

lukewarlow replied:
You were correct I forgot they were dependent on each other in all usages. So I've swapped this for a 4 value enum.

darinadler replied:
Maybe obvious and maybe you already changed things, but: for getters we can still have as many boolean ones as we like to make the code clear at the site where it’s used, whereas the setters and data members can stick to the enumeration. It may turn out that a getter that returns the enumeration might not ever be needed.

lukewarlow replied:
So I've used the enum at the getter places too because I found it to be clearer to read at the call site. Happy to change to boolean getters if that's prefered.

darinadler replied:
If you think it’s clearer at the call sites, we should keep it that way. That’s what matters.

I am surprised to hear that you think the enumeration is clearer than a boolean getter, though; did you actually consider both?

lukewarlow replied:
Yeah I had it as the booleans originally and changed it to just evaluate at the callsite with the enum values. Naming them is difficult given the different states, with the enum it just makes it clear which state I'm checking.

I'm happy to defer if others disagree though. I can be persuaded either way.

darinadler commented on Source/JavaScriptCore/runtime/JSGlobalObject.h:
```diff
@@ -1077,23 +1079,30 @@ class JSGlobalObject : public JSSegmentedVariableObject {
     {
         m_requiresTrustedTypes = required;
+        m_requiresTrustedTypesEnforced = enforced;
     }
```
The old version of this function was easy to understand at the call site. The new one is not. It’s not at all clear what `false, false` means. Can we come up with something where it’s easier to read and understand use of this? Two separate functions would be easier to understand, like `setEnforcesTrustedTypesRequirement`, or could pass in an enumeration with three values for _not required_, _required but not enforced_, _required and enforced_.

darinadler commented on Source/WebCore/bindings/js/ScriptController.h:
```diff
@@ -130,8 +130,9 @@ class ScriptController final : public CanMakeWeakPtr<ScriptController>, public C
+    void setTrustedEvalEnabled(bool);
     void setWebAssemblyEnabled(bool, const String& errorMessage = String());
-    void setRequiresTrustedTypes(bool);
+    void setRequiresTrustedTypes(bool, bool);
```
Same problem here, but it’s even worse because this one we can’t even understand in the header here. We need names for these two `bool` arguments, but really we need a different clearer idiom.

darinadler commented on Source/WebCore/dom/Document.h:
```diff
@@ -862,7 +862,8 @@ class Document
     void disableEval(const String& errorMessage) final;
     void disableWebAssembly(const String& errorMessage) final;
-    void setRequiresTrustedTypes(bool required) final;
+    void setRequiresTrustedTypes(bool required, bool enforced) final;
```
Same pattern again, with some of the same issues.

darinadler commented on Source/WebCore/dom/EmptyScriptExecutionContext.h:
```diff
@@ -62,7 +62,8 @@ class EmptyScriptExecutionContext final : public RefCounted<EmptyScriptExecution
     void disableWebAssembly(const String&) final { };
-    void setRequiresTrustedTypes(bool) final { };
+    void setRequiresTrustedTypes(bool, bool) final { };
+    void setTrustedEvalEnabled(bool) final { };
```
These semicolons aren’t needed. Should remove the existing ones too.

darinadler commented on Source/WebCore/dom/ScriptExecutionContext.h:
```diff
@@ -142,7 +142,8 @@ class ScriptExecutionContext : public SecurityContext, public TimerAlignment {
     virtual void disableEval(const String& errorMessage) = 0;
     virtual void disableWebAssembly(const String& errorMessage) = 0;
-    virtual void setRequiresTrustedTypes(bool required) = 0;
+    virtual void setRequiresTrustedTypes(bool, bool) = 0;
```
Same problem here, but it’s even worse because this one we can’t even understand in the header here. We need names for these two `bool` arguments, but really we need a different clearer idiom.

darinadler commented on Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.cpp:
```diff
@@ -154,7 +159,10 @@ std::unique_ptr<ContentSecurityPolicyDirectiveList> ContentSecurityPolicyDirecti
 
     if (!checkEval(directives->operativeDirective(directives->m_scriptSrc.get(), ContentSecurityPolicyDirectiveNames::scriptSrc)))
-        directives->setEvalDisabledErrorMessage(makeString("Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive: \""_s, directives->operativeDirective(directives->m_scriptSrc.get(), ContentSecurityPolicyDirectiveNames::scriptSrc)->text(), "\".\n"_s));
+        directives->setEvalDisabledErrorMessage(makeString("Refused to evaluate a string as JavaScript because 'unsafe-eval' or 'trusted-types-eval' is not an allowed source of script in the following Content Security Policy directive: \""_s, directives->operativeDirective(directives->m_scriptSrc.get(), ContentSecurityPolicyDirectiveNames::scriptSrc)->text(), "\".\n"_s));
```
Does this message have to be vague? Could it cite the particular source of script the code attempted to use instead of always listing both?

lukewarlow replied:
Well it's the lack of a source that's the issue, so I don't know which they tried to use, because they haven't used either.

justinmichaud commented on Source/JavaScriptCore/debugger/DebuggerEvalEnabler.h:
```diff
@@ -64,8 +63,7 @@ class DebuggerEvalEnabler {
                 globalObject->setEvalEnabled(false, globalObject->evalDisabledErrorMessage());
-            if (m_trustedTypesWereRequired)
-                globalObject->setRequiresTrustedTypes(true);
+            globalObject->setRequiresTrustedTypes(m_requiresTrustedTypes);
```
This is never set, right?

lukewarlow replied:
Oh yeah that's a good catch.

rreno commented on Source/JavaScriptCore/runtime/JSGlobalObject.h:
```diff
@@ -604,6 +610,8 @@ class JSGlobalObject : public JSSegmentedVariableObject {
     std::optional<unsigned> m_stackTraceLimit;
     Weak<FunctionExecutable> m_executableForCachedFunctionExecutableForFunctionConstructor;
 
+    RequiresTrustedTypes m_requiresTrustedTypes { RequiresTrustedTypes::Enforced };
```
Why do we default to enforcing when the feature is off by default in UnifiedWebPreferences?

lukewarlow replied:
The feature is (recently) on by default fwiw. I don't think this default makes a difference either way, but just to be safe I'll set it to NotRequired

kmiller68 commented on Source/JavaScriptCore/runtime/JSGlobalObject.h:
```diff
@@ -140,6 +140,13 @@ enum class CompilationType {
     IndirectEval,
 };
 
+enum class RequiresTrustedTypes {
```
I think the rest of the JSC code would be easier to read if this was `TrustedTypesEnforcement` and maybe `NotRequired` should be `None`

kmiller68 commented on Source/JavaScriptCore/runtime/JSGlobalObject.h:
```diff
@@ -604,6 +610,8 @@ class JSGlobalObject : public JSSegmentedVariableObject {
     Weak<FunctionExecutable> m_executableForCachedFunctionExecutableForFunctionConstructor;
 
+    RequiresTrustedTypes m_requiresTrustedTypes { RequiresTrustedTypes::NotRequired };
+
```
Ditto on calling this `m_trustedTypesEnforcement`

kmiller68 commented on Source/JavaScriptCore/runtime/JSGlobalObject.h:
```diff
@@ -1091,7 +1099,7 @@ class JSGlobalObject : public JSSegmentedVariableObject {
     bool evalEnabled() const { return m_evalEnabled; }
     bool webAssemblyEnabled() const { return m_webAssemblyEnabled; }
-    bool requiresTrustedTypes() const { return m_requiresTrustedTypes; }
+    RequiresTrustedTypes requiresTrustedTypes() const { return m_requiresTrustedTypes; }
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/runtime/JSGlobalObject.h:
```diff
@@ -1104,7 +1112,7 @@ class JSGlobalObject : public JSSegmentedVariableObject {
         m_webAssemblyDisabledErrorMessage = errorMessage;
     }
-    void setRequiresTrustedTypes(bool required)
+    void setRequiresTrustedTypes(RequiresTrustedTypes required)
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/runtime/DirectEvalExecutable.cpp:
```diff
@@ -40,7 +40,7 @@ DirectEvalExecutable* DirectEvalExecutable::create(JSGlobalObject* globalObject,
     auto scope = DECLARE_THROW_SCOPE(vm);
 
-    if (!globalObject->evalEnabled()) {
+    if (!globalObject->evalEnabled() && !(Options::useTrustedTypes() && globalObject->requiresTrustedTypes() == RequiresTrustedTypes::EnforcedWithEvalEnabled)) {
```
Why do we need the `Options::useTrustedTypes()` guard here and other places? Why are we setting the enforcement level if `Options::useTrustedTypes() == false`? These would also generally be a lot easier to read if we pushed the negation into the boolean logic, IMO. So `&& (!Options::useTrustedTypes || globalObject->trustedTypesEnforcement() != TrustedTypesEnforcement:: EnforcedWithEvalEnabled)`

lukewarlow replied:
Good point I've added the flag check into setTrustedTypesEnforcement in JSGlobalObject.

kmiller68 reviewed (changes requested):
See previous comments (forgot to click the button... whoops).

lukewarlow committed:
Implement trusted-types-eval CSP script-src keyword
https://bugs.webkit.org/show_bug.cgi?id=285603

Reviewed by Ryan Reno.

Add a new script-src keyword that allows evaluating strings as JS,
similar to unsafe-eval but with the added requirement that trusted-types
must also be enabled and enforced.

* LayoutTests/http/tests/security/contentSecurityPolicy/block-eval-onload-in-nested-about-blank-iframe-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/eval-blocked-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/eval-blocked-in-about-blank-iframe-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/eval-blocked-in-external-script-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/eval-blocked-in-subframe-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/eval-scripts-setInterval-blocked-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/eval-scripts-setTimeout-blocked-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/function-constructor-blocked-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/module-eval-blocked-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/module-eval-blocked-in-external-script-expected.txt:
* LayoutTests/http/tests/security/contentSecurityPolicy/worker-blob-inherits-csp-blocks-eval-expected.txt:
* LayoutTests/http/tests/security/isolatedWorld/image-load-should-not-bypass-main-world-csp-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/script-src/script-src-trusted_types_eval_with_report_only_require_trusted_types_eval-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/script-src/script-src-trusted_types_eval_with_report_only_require_trusted_types_eval.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/script-src/script-src-trusted_types_eval_with_report_only_require_trusted_types_eval.html.headers: Added.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/script-src/script-src-trusted_types_eval_with_require_trusted_types_eval-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/script-src/script-src-trusted_types_eval_with_require_trusted_types_eval.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/script-src/script-src-trusted_types_eval_with_require_trusted_types_eval.html.headers: Added.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/script-src/script-src-trusted_types_eval_without_require_trusted_types_eval-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/script-src/script-src-trusted_types_eval_without_require_trusted_types_eval.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/script-src/script-src-trusted_types_eval_without_require_trusted_types_eval.html.headers: Added.
* LayoutTests/security/contentSecurityPolicy/worker-inherits-blocks-eval-expected.txt:
* Source/JavaScriptCore/debugger/DebuggerEvalEnabler.h:
(JSC::DebuggerEvalEnabler::DebuggerEvalEnabler):
(JSC::DebuggerEvalEnabler::~DebuggerEvalEnabler):
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::eval):
* Source/JavaScriptCore/runtime/DirectEvalExecutable.cpp:
(JSC::DirectEvalExecutable::create):
* Source/JavaScriptCore/runtime/FunctionConstructor.cpp:
(JSC::constructFunction):
* Source/JavaScriptCore/runtime/IndirectEvalExecutable.cpp:
(JSC::IndirectEvalExecutable::createImpl):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::trustedTypesEnforcement const):
(JSC::JSGlobalObject::setTrustedTypesEnforcement):
(JSC::JSGlobalObject::requiresTrustedTypes const): Deleted.
(JSC::JSGlobalObject::setRequiresTrustedTypes): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/WebCore/bindings/js/ScriptController.cpp:
(WebCore::ScriptController::setTrustedTypesEnforcement):
(WebCore::ScriptController::setRequiresTrustedTypes): Deleted.
* Source/WebCore/bindings/js/ScriptController.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setTrustedTypesEnforcement):
(WebCore::Document::setRequiresTrustedTypes): Deleted.
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/EmptyScriptExecutionContext.h:
* Source/WebCore/dom/ScriptExecutionContext.h:
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::ContentSecurityPolicy::didCreateWindowProxy const):
(WebCore::ContentSecurityPolicy::applyPolicyToScriptExecutionContext):
(WebCore::ContentSecurityPolicy::allowEval const):
(WebCore::ContentSecurityPolicy::requireTrustedTypesForSinkGroup const):
* Source/WebCore/page/csp/ContentSecurityPolicy.h:
* Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.cpp:
(WebCore::checkTrustedEval):
(WebCore::ContentSecurityPolicyDirectiveList::create):
* Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.h:
(WebCore::ContentSecurityPolicyDirectiveList::trustedEvalEnabled const):
(WebCore::ContentSecurityPolicyDirectiveList::setTrustedEvalEnabled):
* Source/WebCore/page/csp/ContentSecurityPolicySourceList.cpp:
(WebCore::ContentSecurityPolicySourceList::parseSource):
* Source/WebCore/page/csp/ContentSecurityPolicySourceList.h:
(WebCore::ContentSecurityPolicySourceList::allowTrustedEval const):
* Source/WebCore/page/csp/ContentSecurityPolicySourceListDirective.h:
(WebCore::ContentSecurityPolicySourceListDirective::allowTrustedEval const):
* Source/WebCore/workers/WorkerOrWorkletGlobalScope.cpp:
(WebCore::WorkerOrWorkletGlobalScope::setTrustedTypesEnforcement):
(WebCore::WorkerOrWorkletGlobalScope::setRequiresTrustedTypes): Deleted.
* Source/WebCore/workers/WorkerOrWorkletGlobalScope.h:
* Source/WebCore/workers/WorkerOrWorkletScriptController.cpp:
(WebCore::WorkerOrWorkletScriptController::setTrustedTypesEnforcement):
(WebCore::WorkerOrWorkletScriptController::setRequiresTrustedTypes): Deleted.
* Source/WebCore/workers/WorkerOrWorkletScriptController.h:

Canonical link: https://commits.webkit.org/292217@main
